### PR TITLE
refactor: Clarify what's happening in the hardcoded compactor components

### PR DESCRIPTION
**Pull request by carols10cents**
I started working on adding selection of partitions for cold compaction, and I again found it difficult to follow what was happening in `hardcoded_components`: there were pieces of code that were related but had different code between them, there were parts that depended on earlier parts but not all of them, there were functions for some parts but not others, etc. Overall, it required reading the whole function and thinking carefully before changing anything.

To make this better, I extracted each piece into its own function, which solves these problems for me:

- It's now much clearer which parts are related and which parts are independent. For example, the `partitions_source`, `commit`, and `partition_done_sink` components are intertwined, which is made clear by creating them all in one function.
- It's now clearer which parts make use of other parts created earlier. When possible, pieces are created within each function and only the final result that ends up in `Components` is exposed in `hardcoded_components`. An exception is the `partitions_source`, which is used as input to creating the `partition_stream`. I chose to leave inline those components that don't even depend on `config` for their creation, such as `round_split`. 
- When I want to modify a particular piece, I only need to understand one or two relevant functions, not all of the code that goes into creating the `Components`.

There's definitely more that could be done here, such as making these functions into methods on some of the component objects perhaps, passing in only the relevant fields of the `config`, splitting up `make_partitions_source_commit_partition_sink` more, reducing duplication when wrapping in logging/metrics objects, but I think this is a good step forward.

I highly recommend reviewing commit-by-commit; each change should be fairly small and easy to review (mostly just moving code into a function) and each commit passes the `compactor2` tests.

### CODE REVIEW

File: compactor2/src/components/hardcoded.rs [hunk 1]
```diff
@@ -138,86 +107,12 @@ pub fn hardcoded_components(config: &Config) -> Arc<Components> {
...
-    };
-
     let scratchpad_store_output = if config.shadow_mode {
         Arc::new(IgnoreWrites::new(Arc::new(InMemory::new())))
     } else {
         Arc::clone(config.parquet_store_real.object_store())
     };
 
-    let (partitions_source, partition_done_sink) =
-        unique_partitions(partitions_source, partition_done_sink, 1);
```
Comment (carols10cents) [inline compactor2/src/components/hardcoded.rs:175]: For example, I found this very confusing that the `partition_filters` and `scratchpad_store_output` were built between starting to build the `partitions_source` and then continuing to change `partitions_source` here, but `partition_filters` and `scratchpad_store_output` don't affect `partitions_source` at all.

File: compactor2/src/components/hardcoded.rs [hunk 2]
```diff
@@ -363,11 +340,27 @@ fn make_partition_stream(
 }
 
 // Conditions to compact this partittion
-fn make_partition_filters(config: &Config) -> Vec<Arc<dyn PartitionFilter>> {
-    // (Has-L0) OR            -- to avoid overlaped files
+fn make_partition_filter(config: &Config) -> Arc<dyn PartitionFilter> {
```
Comment (carols10cents) [inline compactor2/src/components/hardcoded.rs:283]: This function already existed but only built a sub-part of `partition_filter`; I'm not sure why different parts were in different places but now the whole construction is in this one function.

File: compactor2/src/components/hardcoded.rs [hunk 3]
```diff
@@ -305,7 +308,7 @@ fn make_partitions_source_commit_partition_sink(
         ))
     };
 
-    (partitions_source, Arc::new(commit), partition_done_sink)
+    (partitions_source, commit, partition_done_sink)
```
Comment (carols10cents) [inline compactor2/src/components/hardcoded.rs:241]: Here's where the `Arc::new` is no longer needed

**Comment by carols10cents** [comment]
Switching to draft while I look into the test failure

File: compactor2/src/components/hardcoded.rs [hunk 4]
```diff
@@ -382,5 +375,15 @@ fn make_partition_filters(config: &Config) -> Vec<Arc<dyn PartitionFilter>> {
...
+
+    let partition_continue_conditions = "continue_conditions";
+    Arc::new(LoggingPartitionFilterWrapper::new(
+        MetricsPartitionFilterWrapper::new(
+            AndPartitionFilter::new(partition_filters),
+            &config.metric_registry,
+            partition_continue_conditions,
+        ),
+        partition_continue_conditions,
+    ))
```
Comment (NGA-TRAN) [inline compactor2/src/components/hardcoded.rs:327]: This is the comment for the commit "Move all partition filter creation into the function for that purpose"

The refactor looks good but I suggest inside `make_partition_filters` make 2 sub-functions, one for the new ones you just moved in and one for the available ones.

For you to name those 2 sub-functions:
- The one you just moved in is something like `these are exceptional cases we do not want to deal with no matter what`
- The existing one is: `the conditions to continue compacting`

By separating them, we will make our code cleaner and I bet when you work on the conditions for cold compaction, you will only need to modify to the second one

**Review by NGA-TRAN — COMMENTED** [review]
I have reviewed commit by commit. They all make sense and thanks for doing this @carols10cents . At first the hardcoded_components is very small but then we kept adding more until it is hard to follow if you are new. This is needed.

I have only one suggestion to add 2 sub-functions for  partition filters to make it  clear about 2 kinds of filters and easy for future added conditions for cold compaction

File: iox_catalog/src/postgres.rs [hunk 5]
```diff
@@ -1532,18 +1532,29 @@ WHERE id = $2;
...
+            {}
             "#,
-        )
-        .bind(recent_time) // $1
-        .fetch_all(&mut self.inner)
-        .await
-        .map_err(|e| Error::SqlxError { source: e })
+            maximum_time
+                .map(|_| "AND p.new_file_at < $2")
+                .unwrap_or_default()
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply iox_catalog/src/postgres.rs:1549]: Yes, the `map` returns an `Option<&str>`, and the default of `&str` is `""`, so there will be an empty string in the `{}` if `maximum_time` is `None`.

I actually didn't mean to include this commit in this PR, it's setting up for the cold compaction selection, but if you'll approve it now I'll take it 😅 I will add the extra test case you suggested!

[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply iox_catalog/src/postgres.rs:1549]: Yeah, I was aware this is prepared for cold compaction. I saw the tests wok well and it is good to go

**Review by NGA-TRAN — APPROVED** [review]
They look good.
